File: python/atelier_bench/repositories/technician_repository.py
"""TechnicianRepository: pins Repository[int, Technician]; the generic Protocol is satisfied structurally."""

from __future__ import annotations

from collections.abc import Sequence

from atelier_core import Technician
from atelier_core.contracts.repository import Repository


class TechnicianRepository:
    """Satisfies `Repository[int, Technician]` without naming it."""

    def __init__(self, records: Sequence[Technician] = ()) -> None:
        self.records = list(records)

    def find(self, key: int) -> Technician | None:
        return next((record for record in self.records if record.id == key), None)

    def all(self) -> Sequence[Technician]:
        return self.records

    def count(self) -> int:
        return len(self.records)

    def available(self) -> list[Technician]:
        return [technician for technician in self.records if technician.next_slot() is not None]


def as_repository(repository: TechnicianRepository) -> Repository[int, Technician]:
    """The only place the Protocol name meets the class: a typed pass-through."""
    return repository

File: go/atelier/repositories/technician_repository.go
// Package repositories carries the 8 repositories. Each satisfies the GENERIC
// interface core/contracts.Repository[T] with its own element type; the only
// textual link is the compile-time assertion at the top of the file.
package repositories

import (
	"atelier.example/lane/core/contracts"
	"atelier.example/lane/core/models"
)

// Compile-time proof: the generic interface instantiated with *models.Technician.
var _ contracts.Repository[*models.Technician] = (*TechnicianRepository)(nil)

// TechnicianRepository holds the technicians in memory.
type TechnicianRepository struct {
	records []*models.Technician
}

// NewTechnicianRepository builds a repository over the given records.
func NewTechnicianRepository(records []*models.Technician) *TechnicianRepository {
	return &TechnicianRepository{records: records}
}

// All returns every record.
func (r *TechnicianRepository) All() []*models.Technician {
	return r.records
}

// Find resolves one record by id.
func (r *TechnicianRepository) Find(id int) (*models.Technician, bool) {
	for _, record := range r.records {
		if record.ID == id {
			return record, true
		}
	}
	var zero *models.Technician
	return zero, false
}

// Count is how many records the repository holds.
func (r *TechnicianRepository) Count() int {
	return len(r.records)
}

// Available is the repository-specific query.
func (r *TechnicianRepository) Available() []*models.Technician {
	found := make([]*models.Technician, 0, len(r.records))
	for _, record := range r.records {
		if record.NextSlot() >= 0 {
			found = append(found, record)
		}
	}
	return found
}
